feat(hesai): software-based precise FoV limit & early publishing

mojomex opened the pull request:
## PR Type

- New Feature

## Description

This PR introduces software-based FoV limits and early publishing for Hesai sensors.

### Software-based FoV
Previously, configuring `cloud_min_angle` and `cloud_max_angle` would set the sensor's hardware filter to those values.
However, the hardware filter is using **raw azimuths**.
This lead to the scan having jagged edges (points missing in some channels, too many points in others).
For certain models, this affected areas of up to ~10deg on both edges of the FoV.

This PR improves upon this in the following way:
1. User sets `cloud_min_angle` and `cloud_max_angle` to the exact FoV they want the output pointcloud to have
2. Nebula retrieves sensor calibration file and computes appropriately over-sized* hardware-FoV
3. Nebula sends hardware-FoV to sensor (or validates it if `setup_sensor == false`)
4. The Nebula decoder cuts the scan at the exact edges defined via the parameters above

### Early publishing
Previously, a scan was published once the packet **after** it was received. This was done because this made the scan completion check easier (could compare 2 azimuths according to some scan completion criterion).
For scans of 360deg, this is not a problem, as there is no time gap between the end of one scan and the beginning of the next.
However, for FoVs < 360deg, this approach increases latency unnecessarily: for an FoV of 180deg and 1 full rotation every 100ms, the additional latency of the finished pointcloud just "sitting there" is 50ms.

This PR improves upon this in the following way:
1. Swap the order of decoding and scan completion check: decode first, then check for completion
2. Check if the current packet completed the current scan, as opposed to whether the current packet is the first belonging to the next scan

### Other

For testing this PR with AT128, I had to fix the LidarRange issue:
* fixes #129 

This was done by checking the sensor model before sending the LidarRange command.

## Review Procedure

Check for the following FoVs
* 0 - 360 deg
* 90 - 270 deg
* 270 - 90 deg
if the pointcloud in Rviz has clean edges at those angles for a 360 deg sensor such as OT128.

Check AT128 in the same way but with FoVs inside [30; 150].

## Remarks

These two improvements have been combined into one PR as they are closely related and would require extra work to function independently.

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [x] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR**: Required to pass before the merge.

mojomex commented:
# :construction: Evaluation

## :green_circle: 360deg rotational sensors

| FoV Setting (min, max) | Resulting pointcloud |
|--------|--------|
| 0, 360 | ![0-360](https://github.com/tier4/nebula/assets/6088931/40eac350-483c-4c56-a953-346195b852be) |
| 90, 270 | ![90-270](https://github.com/tier4/nebula/assets/6088931/fa307697-8a8d-4a40-8c65-279c0c93889b) |
| 270, 90 | ![270-90](https://github.com/tier4/nebula/assets/6088931/ef3f685d-ff42-47fa-87f6-698561cc08fd) |
| 29, 30 | ![29-30](https://github.com/tier4/nebula/assets/6088931/27af3771-0649-4227-9edc-e04658ceaeb5) |
| 360, 0 | ![360-0](https://github.com/tier4/nebula/assets/6088931/892f28d4-44dd-45c9-8f0e-e00a6e09a17d) | 

## :green_circle: AT128 (120deg rotational sensor)

| FoV Setting (min, max) | Resulting pointcloud |
|--------|--------|
| 0, 360 | ![0-360](https://github.com/tier4/nebula/assets/6088931/e938d7c5-ef1e-4977-8c19-359ec4a62237) |
| 30, 150 | ![30-150](https://github.com/tier4/nebula/assets/6088931/13647cab-ab65-433a-b34d-0e9d4d9fbe30) |
| 60, 120 | ![60-120](https://github.com/tier4/nebula/assets/6088931/0f2133fe-a44c-48ec-978d-eea25d704ab2) |
| 89, 90 | ![89-90](https://github.com/tier4/nebula/assets/6088931/90ed7d44-e8f8-485d-8c4e-ed6b01badc80) |

drwnz reviewed (changes requested):
This was tested with 40P with the following issue:
- Setting cut angle out side of the range of FoV start and end, resulted in no error and also pointcloud output.
- Even if that was the case and the "cut" happens in the area outside of the FoV (which is kind of uneccesary but logically possible) and the actual observed point cloud is correct in terms of min and max angles.
- Example: cut 10, min 180, max 360: no warnings, output pointcloud is full 180 degree segment.

mojomex commented on nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/angle_corrector_correction_based.hpp:
```diff
@@ -79,31 +179,62 @@ class AngleCorrectorCorrectionBased : public AngleCorrector<HesaiCorrection>
     auto elevation =
```
Changed to `int32_t`.

mojomex replied:
Addressed in [f4f8bf1](https://github.com/tier4/nebula/pull/173/commits/f4f8bf1f0c3f2a3abe7755641794d1f419c082c1)

mojomex commented on nebula_common/include/nebula_common/hesai/hesai_common.hpp:
```diff
@@ -357,7 +377,7 @@ struct HesaiCorrection : public HesaiCalibrationConfigurationBase
   /// @param azi The precision azimuth in (0.01 / 256) degree unit
   /// @return The azimuth adjustment in 0.01 degree unit
-  int8_t getAzimuthAdjustV3(uint8_t ch, uint32_t azi) const
+  [[nodiscard]] int8_t getAzimuthAdjustV3(uint8_t ch, uint32_t azi) const
```
Agreed :+1:

mojomex replied:
Agreed :+1:

mojomex commented on nebula_common/include/nebula_common/hesai/hesai_common.hpp:
```diff
@@ -369,13 +389,19 @@ struct HesaiCorrection : public HesaiCalibrationConfigurationBase
+  [[nodiscard]] std::tuple<float, float> getFovPadding() const override
+  {
+    // TODO(mojomex): calculate instead of hard-coding
+    return {-5, 5};
```
Addressed in [7272be2](https://github.com/tier4/nebula/pull/173/commits/7272be2936e9f381e66fc3b60b14c81fd3aa2b52)

mojomex commented on nebula_tests/hesai/hesai_ros_scan_cutting_test_main.cpp:
```diff
@@ -0,0 +1,179 @@
+        threshold_angle, drivers::normalize_angle(fov_max_rad, 2 * M_PIf), p.azimuth);
+    }
+
+    bool is_at128 = drivers::SensorModelFromString(hesai_driver_->params_.sensor_model) ==
```
Addressed in [eaa49f6](https://github.com/tier4/nebula/pull/173/commits/eaa49f697029f7a3306b2726782008a86073056a)

mojomex commented:
@drwnz Thank you for the review!

I have addressed your comments and re-tested setting the `cut_angle` parameter to a value outside of the FoV.
Nebula now properly emits an error:
```
[INFO] [hesai_ros_wrapper_node]: OnParameterChange
[ERROR] [hesai_ros_wrapper_node]: Cannot cut scan outside of the FoV.
[WARN] [hesai_ros_wrapper_node]: OnParameterChange aborted: Could not set SensorConfiguration
```

The fix is in this commit: [2e42bd0](https://github.com/tier4/nebula/pull/173/commits/2e42bd045c3c2d5c3124ab7ea15b5ba4b8e94e6e)

drwnz commented on nebula_hw_interfaces/include/nebula_hw_interfaces/nebula_hw_interfaces_hesai/hesai_hw_interface.hpp:
```diff
@@ -331,6 +331,16 @@ class HesaiHwInterface
+   * compensates for the points lost due to the sensor filtering FoV by raw encoder angle.
+   *
+   * @param calibration The calibration file of the sensor
+   * @return Status If the FoV was updated correctly
```
```suggestion
   * @return Status Resulting status of setting the FoV 
```

drwnz commented on nebula_hw_interfaces/src/nebula_hesai_hw_interfaces/hesai_hw_interface.cpp:
```diff
@@ -532,6 +542,32 @@ HesaiLidarRangeAll HesaiHwInterface::GetLidarRange()
+    return Status::SENSOR_CONFIG_ERROR;
+  }
+
+  int cloud_min = sensor_configuration_->cloud_min_angle * 10;
```
Nit: Instead of multiplying by 10 and using 3600, it would be tidier to define decidegrees so it is clear when we are dealing in normal degrees or decidegrees.

mojomex replied:
Agreed. Since a lot of the values come from other places that are in degrees, the `* 10` is almost inevitable.
The cleanest (but for this PR overkill) approach would be proper types like `Degrees`, `DeciDegrees` which can be cast properly. See [std::chrono::duration](https://en.cppreference.com/w/cpp/chrono/duration) for nice unit handling.

drwnz commented on nebula_ros/include/nebula_ros/hesai/hesai_ros_wrapper.hpp:
```diff
@@ -29,17 +26,16 @@
 #include <rclcpp_components/register_node_macro.hpp>
 
-#include "nebula_msgs/msg/nebula_packet.hpp"
+#include <pandar_msgs/msg/pandar_scan.hpp>
```
No need for nebula_packets and nebula_packet? (Ignore if not, I haven't gone through the include tree)

mojomex replied:
Good catch! Re-added in dbd8262

drwnz commented on nebula_ros/include/nebula_ros/hesai/hesai_ros_wrapper.hpp:
```diff
@@ -79,6 +78,23 @@ class HesaiRosWrapper final : public rclcpp::Node
 
+  /// @brief The ROS 2 parameter holding the calibration file path is called differently depending
+  /// on the sensor model. This function returns the correct parameter name given a model.
+  /// @param model A sensor model
```
```suggestion
  /// @param model The sensor model
```

drwnz commented on nebula_ros/schema/sub/lidar_hesai.json:
```diff
@@ -32,6 +32,20 @@
+    "cut_angle": {
+      "type": "number",
+      "minimum": 0,
+      "maximum": 360,
```
[Future work] in the documentation it may be worth explaining why the cut angle can go to 360, but sync angle can not. Sync angle is an integer and cut angle is a float would be sufficient, it's just a little confusing how the sensor and driver handles 360 vs 0.

drwnz commented on nebula_ros/src/hesai/decoder_wrapper.cpp:
```diff
@@ -255,6 +131,12 @@ void HesaiDecoderWrapper::ProcessCloudPacket(
   cloud_watchdog_->update();
 
+  auto sys_pc_time_diff = parent_node_.now().seconds() - std::get<1>(pointcloud_ts);
+  // RCLCPP_INFO_STREAM(
```
Leave commented code for easy debug later, or delete if no longer needed?

mojomex replied:
Removed in b54891e.

drwnz commented on nebula_ros/src/hesai/hesai_ros_wrapper.cpp:
```diff
@@ -88,9 +113,21 @@ nebula::Status HesaiRosWrapper::DeclareAndGetSensorConfigParams()
     descriptor.floating_point_range = float_range(0, 360, 0.01);
-    config.scan_phase = declare_parameter<double>("scan_phase", descriptor);
+    descriptor.description =
+      "At which angle to start a new scan. Cannot be equal to the start angle in a non-360 deg "
```
```suggestion
      "Angle at which to cut the scan. Cannot be equal to the start angle in a non-360 deg "
```

drwnz commented on nebula_ros/src/hesai/hesai_ros_wrapper.cpp:
```diff
@@ -292,6 +380,22 @@ rcl_interfaces::msg::SetParametersResult HesaiRosWrapper::OnParameterChange(
+    auto status = hw_interface_wrapper_->HwInterface()->checkAndSetLidarRange(*new_calibration_ptr);
+    if (status != Status::OK) {
+      RCLCPP_ERROR_STREAM(
+        get_logger(), "Could not change FoV but config already changed: " << status);
```
This error message is a little contradictory. Is it because everything has been set except range which failed?
In which case, something like "Sensor configuration updated, but FoV could not be set" might be better.

mojomex replied:
Re-worded in 7d4effa

drwnz reviewed (commented):
I left a few more comments, nothing urgent.
I haven't gone through the tests except for verifying that they passed.

mojomex commented on nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/angle_corrector_correction_based.hpp:
```diff
@@ -53,9 +78,40 @@ class AngleCorrectorCorrectionBased : public AngleCorrector<HesaiCorrection>
     return field;
   }
 
+  uint32_t all_channels(uint32_t azi, double threshold, bool any, bool eq_ok)
```
Addressed in [63233e7](https://github.com/tier4/nebula/pull/173/commits/63233e7686298ceb40fba031688d5d3bfbcad2a1)

mojomex commented on nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/hesai_decoder.hpp:
```diff
@@ -154,17 +172,38 @@ class HesaiDecoder : public HesaiScanDecoder
-        NebulaPoint point;
+        CorrectedAngleData corrected_angle_data =
+          angle_corrector_.getCorrectedAngleData(raw_azimuth, channel_id);
+        auto azimuth = corrected_angle_data.azimuth_rad;
```
Addressed in 98808d4

mojomex commented on nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/hesai_decoder.hpp:
```diff
@@ -243,42 +273,48 @@ class HesaiDecoder : public HesaiScanDecoder
+      if (angle_corrector_.passedEmitAngle(last_azimuth_, block_azimuth)) {
         std::swap(decode_pc_, output_pc_);
-        decode_pc_->clear();
+        std::swap(decode_scan_timestamp_ns_, output_scan_timestamp_ns_);
```
Added clarifications in 0bc2db9.
I agree the definitions are not strictly `decode` and `output` anymore (`deocde` and `overflow` from before also do not quite fit anymore).

However,

* the `decode` cloud is always used for decoding, except when cutting within the FoV and scans overlap
* in that case, the `output` and `decode` clouds are decoded to simultaneously in the overlap region
* the `output` cloud is always the one that is output when a scan is completed.

As I can't think of any better naming right now, I'd suggest leaving it as-is until we think of something better :bow:

drwnz reviewed (approved):
Ready to go, great work!
